pkg/ddl/notifier/subscribe.go

```diff
@@ -0,0 +1,208 @@
+// ID occupies a bit in a BIGINT column, so at most we can only have 64 IDs. To
+// avoid duplicate IDs, all IDs should be defined in below declaration.
+type HandlerID int
+
+const (
+	// TestHandlerID is used for testing only.
+	TestHandlerID HandlerID = iota
+)
+
+// RegisterHandler must be called with an exclusive and fixed HandlerID for each
+// handler to register the handler. Illegal ID will panic. RegisterHandler should
+// not be called after the global ddlNotifier is started.
+//
+// RegisterHandler is not concurrency-safe.
+func RegisterHandler(id HandlerID, handler SchemaChangeHandler) {
+	intID := int(id)
+	if intID < 0 || intID >= 64 {
+		panic(fmt.Sprintf("illegal HandlerID: %d", id))
+	}
+
+	if _, ok := globalDDLNotifier.handlers[id]; ok {
+		panic(fmt.Sprintf("HandlerID %d already registered", id))
+	}
+	globalDDLNotifier.handlers[id] = handler
+}
+
+type ddlNotifier struct {
+	ownedSCtx    sessionctx.Context
+	store        Store
+	handlers     map[HandlerID]SchemaChangeHandler
+	pollInterval time.Duration
+}
+
+var globalDDLNotifier *ddlNotifier
+
+// InitDDLNotifier initializes the global ddlNotifier. It should be called only
+// once and before any RegisterHandler call. The ownership of the sctx is passed
+// to the ddlNotifier.
+func InitDDLNotifier(
+	sctx sessionctx.Context,
+	store Store,
+	pollInterval time.Duration,
+) {
+	globalDDLNotifier = &ddlNotifier{
+		ownedSCtx:    sctx,
+		store:        store,
+		handlers:     make(map[HandlerID]SchemaChangeHandler),
+		pollInterval: pollInterval,
+	}
+}
+
+// ResetDDLNotifier is used for testing only.
+func ResetDDLNotifier() { globalDDLNotifier = nil }
+
+// StartDDLNotifier starts the global ddlNotifier. It will block until the
+// context is canceled.
+func StartDDLNotifier(ctx context.Context) {
+	globalDDLNotifier.Start(ctx)
+}
+
+// Start starts the ddlNotifier. It will block until the context is canceled.
+func (n *ddlNotifier) Start(ctx context.Context) {
+	ctx = kv.WithInternalSourceType(ctx, kv.InternalDDLNotifier)
+	ticker := time.NewTicker(n.pollInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			if err := n.processEvents(ctx); err != nil {
+				logutil.Logger(ctx).Error("Error processing events", zap.Error(err))
+			}
+		}
+	}
+}
+
+func (n *ddlNotifier) processEvents(ctx context.Context) error {
+	events, err := n.store.List(ctx, sess.NewSession(n.ownedSCtx))
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// we should ensure deliver order of events to a handler, so if a handler returns
+	// error for previous events it should not receive later events.
+	skipHandlers := make(map[HandlerID]struct{})
+	for _, event := range events {
+		for handlerID, handler := range n.handlers {
+			if _, ok := skipHandlers[handlerID]; ok {
+				continue
+			}
+			if err2 := n.processEventForHandler(ctx, event, handlerID, handler); err2 != nil {
+				skipHandlers[handlerID] = struct{}{}
```

Comment:
Do you mean that if one event handling fails, we should skip all events for this handler?

Reply:
I guess it makes sense for 99% of handlers.

Reply:
We must do this because the notifier should guarantee the delivery order.

Reply:
oh, I forgot that.

Reply:
Maybe we can add some comments here?

Reply:
I have added in line 142